geopandas/clip.py

```diff
@@ -0,0 +1,229 @@
+        | (clipped.geometry.type == "LineString")
+    ]
+    line_diss = lines.dissolve(by=[lines.index]).drop(columns="level_1")
+
+    polys = clipped[clipped.geometry.type == "Polygon"]
+    poly_diss = polys.dissolve(by=[polys.index]).drop(columns="level_1")
+
+    return gpd.GeoDataFrame(
+        pd.concat([poly_diss, line_diss], ignore_index=True)
+    )
+
+
+def clip_shp(shp, clip_obj):
+    """Clip points, lines, or polygon geometries to the clip_obj extent.
+
+    Both layers must be in the same Coordinate Reference System (CRS) and will
+    be clipped to the full extent of the clip object.
+
+    If there are multiple polygons in clip_obj,
+    data from shp will be clipped to the total boundary of
+    all polygons in clip_obj.
+
+    Parameters
+    ----------
+    shp : GeoDataFrame
+          Vector layer (point, line, polygon) to be clipped to clip_obj.
+    clip_obj : GeoDataFrame
+          Polygon vector layer used to clip shp.
+          The clip_obj's geometry is dissolved into one geometric feature
+          and intersected with shp.
+
+    Returns
+    -------
+    GeoDataFrame
+         Vector data (points, lines, polygons) from shp clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clipping points (glacier locations in the state of Colorado) with
+    a polygon (the boundary of Rocky Mountain National Park):
+
+        >>> import geopandas as gpd
+        >>> import earthpy.clip as cl
+        >>> from earthpy.io import path_to_example
+        >>> rmnp = gpd.read_file(path_to_example('rmnp.shp'))
+        >>> glaciers = gpd.read_file(path_to_example('colorado-glaciers.geojson'))
+        >>> glaciers.shape
+        (134, 2)
+        >>> rmnp_glaciers = cl.clip_shp(glaciers, rmnp)
+        >>> rmnp_glaciers.shape
+        (36, 2)
+
+    Clipping a line (the Continental Divide Trail) with a
+    polygon (the boundary of Rocky Mountain National Park):
+
+        >>> cdt = gpd.read_file(path_to_example('continental-div-trail.geojson'))
+        >>> rmnp_cdt_section = cl.clip_shp(cdt, rmnp)
+        >>> cdt['geometry'].length > rmnp_cdt_section['geometry'].length
+        0    True
+        dtype: bool
+
+    Clipping a polygon (Colorado counties) with another polygon
+    (the boundary of Rocky Mountain National Park):
+
+        >>> counties = gpd.read_file(path_to_example('colorado-counties.geojson'))
+        >>> counties.shape
+        (64, 13)
+        >>> rmnp_counties = cl.clip_shp(counties, rmnp)
+        >>> rmnp_counties.shape
+        (4, 13)
+    """
+    try:
+        shp.geometry
+        clip_obj.geometry
+    except AttributeError:
+        raise AttributeError(
+            "Please make sure that your input and clip GeoDataFrames have a"
+            " valid geometry column"
+        )
+
+    if not any(shp.intersects(clip_obj.unary_union)):
+        raise ValueError("Shape and crop extent do not overlap.")
+
+    if any(shp.geometry.type == "MultiPoint"):
```

Comment:
we struggled with this too @martinfleis . here is my question for you. 
the use case we were considering was a shapefile import which will always (i think??) have one geometry type... unless esri has changed things. but ... i could see this being problematic for formats ? 

Can you tell me what type of format might have multiple geometry types natively? it would better help me understand the use cases here!

Reply:
In geopandas we generally do not expect shapefile import as the source of the data can be various. I personally try to avoid shp as much as I can :).

GeoDaraFrame can contain multiple geometry types. You don't need to import them, they can be a result of some kind of manipulation within Geopandas (like intersections of Polygons resulting in LineString).